plugins/traction_innkeeper/traction_innkeeper/v1_0/tenant/routes.py

```diff
@@ -78,6 +80,33 @@ class TenantLedgerIdConfigSchema(OpenAPISchema):
         required=True,
     )
 
+@web.middleware
+async def setup_tenant_context(request: web.Request, handler):
+    """Middle ware to extract tenant_id and provide it to log formatter
+
+    This middleware is appended to the app middlewares and therefore runs
+    last. At this point the wallet_id has been extracted from a previous
+    middleware function and is used to query the tenant record.
+
+    Note: this function is very similar to tenant_self. Can we combine this logic?
+    """
+    context: AdminRequestContext = request["context"]
+    wallet_id, tenant_id = context.profile.settings.get("wallet.id"), None
+
+    if wallet_id:
+        mgr = context.inject(TenantManager)
+        profile = mgr.profile
+
+        async with profile.session() as session:
+            # Tenant records must always be fetched by their wallet id.
+            rec = await TenantRecord.query_by_wallet_id(session, wallet_id)
+            LOGGER.debug(rec)
+            tenant_id = rec.tenant_id
+
+    log_records_inject(tenant_id)
```

Comment:
Running this on every request that goes through the middleware will cause the custom factory to create a new object to handle that specific request, that is used once and then replaced by a - identical in functionality - instance. if we were running out of resources in the PR pods this might be why.

Is there a way to set-up the log factory only *once* when the services start, rather than re-creating it for every single request? A singleton pattern comes to mind (i.e.: https://python-patterns.guide/gang-of-four/singleton), but there may be a better "pythonic" way to do this.

Reply:
Have you seen any pods crashing since this most recent change? I'm not sure how else to inject the `tenant_id` another way.

Reply:
I have not (yet) put the PR instance through some load - will try later and report back.

Reply:
Will do further performance assessment in dev and see if/how to tackle this.